packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,351 @@ import {
-    if (name.labels.length !== 2) return false;
+    if (existingRegistration) {
+      if (
+        existingRegistration.registrationTimestamp &&
+        atTimestamp.time < existingRegistration.registrationTimestamp.time
+      ) {
+        // if the renewal is requested before the registration, we can't renew it
+        // TODO: check for possible off-by-1 errors in the logic above
+        return false;
+      }
 
-    // last label must be "eth"
-    if (name.labels[1] !== ETH_TLD) return false;
+      const releaseTimestamp = getDomainReleaseTimestamp(existingRegistration);
 
-    // first label must be of sufficient length
-    return charCount(name.labels[0]) >= MIN_ETH_REGISTRABLE_LABEL_LENGTH;
+      if (releaseTimestamp && releaseTimestamp.time > atTimestamp.time) {
+        // if the name is released, we can't renew it anymore
+        // TODO: check for possible off-by-1 errors in the logic above
+        return false;
+      }
+    }
+
+    if (duration.seconds < 1n) {
+      // always need to renew for at least 1 second
+      return false;
+    }
+
+    return true;
   }
-}
 
-export const WRAPPED_MAINNET_ETH_REGISTRAR = new NameWrapper(
-  WRAPPED_MAINNET_ETH_REGISTRAR_CONTRACT,
-);
-export const UNWRAPPED_MAINNET_ETH_REGISTRAR =
-  new ClassicETHRegistrarController(UNWRAPPED_MAINNET_ETH_REGISTRAR_CONTRACT);
-
-export const KNOWN_REGISTRARS = [
-  WRAPPED_MAINNET_ETH_REGISTRAR,
-  UNWRAPPED_MAINNET_ETH_REGISTRAR,
-];
-
-export function getPotentialKnownRegistrars(
-  name: ENSName,
-  chain: ChainId,
-  isWrapped: boolean,
-): Registrar[] {
-  return KNOWN_REGISTRARS.filter(
-    (registrar) =>
-      registrar.contract.chain.chainId === chain.chainId &&
-      registrar.isClaimable(name, isWrapped),
-  );
-}
+  public getRenewalPriceQuote(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration: Registration,
+  ): RenewalPriceQuote {
+    if (!this.canRenew(name, atTimestamp, duration, existingRegistration)) {
+      throw new Error(
+        `Cannot build renewal price quote for name: "${name.name}".`,
+      );
+    }
 
-/**
- * Identifies the registrar for the provided name, if known.
- *
- * @param name the name to evaluate.
- * @param chainId the id of the chain the name is managed on.
- * @param isWrapped if the name is wrapped or not.
- * @returns the requested registrar
- */
-export function getKnownRegistrar(
-  name: ENSName,
-  chain: ChainId,
-  isWrapped: boolean,
-): Registrar {
-  const registrars = getPotentialKnownRegistrars(name, chain, isWrapped);
-  if (registrars.length > 1) {
-    throw new Error(
-      `Multiple potential registrars found for name: "${name.name}" on chainId: ${chain.chainId} when isWrapped: ${isWrapped}`,
+    if (!existingRegistration.expirationTimestamp) {
+      throw new Error(`Invariant violation`); // TODO: refine message... just making the type system happy.
+    }
+
+    // TODO: review for possible off-by-1 errors
```

Comment:
@lightwalker-eth can you extend on off-by-1 spec?

Reply:
We need to see how ENS smart contracts actually work. Ex: Actual expiration time? 1 second earlier? 1 second later?

Reply:
Our logic seems to implement the same as below:

<img width="924" alt="Captura de Tela 2024-08-06 às 20 29 09" src="https://github.com/user-attachments/assets/f1a758fe-35f5-467a-be64-98ed7605cb79">

When summing `expiries[id] + duration` to determine the new timestamp.